### PR TITLE
[plugin/feedback-entity] Improvements to Feedback Plugin UI

### PR DESCRIPTION
**Pull request by nikolarAutodesk**
## Hey, I just made a Pull Request!

Our team uses this awesome component and found some things that can improve 🎉

In the Feedback Response Table, there was a floating [Chip](https://github.com/backstage/backstage/blob/7ee4fdcb5485d1941ccbe0e90c27b8fe11260e27/plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx#L84) when a user submits an empty response. 
We also noticed a typo in `FeedbackResponseDialog`. 

While making these fixes, it took a while to get my local set up so I added a log message that would have been helpful to see. 

**Changes**:
- Fix state in `FeedbackResponseTable` to not render an empty `Chip` 
- Fix typo in `FeedbackResponseDialog` (Choose all that ~~applies~~ --> Choose all that apply)
- In `plugins/entity-feedback-backend/src/service/router.ts` I also updated the 400 response to return an explanation. This was rather confusing and took a while to figure out why this was not working, it would have been much easier if I had response text. I looked up the "blessed" way to do this [here](https://github.com/backstage/backstage/blob/master/SECURITY.md#express-responses)

Screenshot 2024-01-23 at 2 11 54 PM
Screenshot 2024-02-21 at 11 03 55 AM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/perfect-shoes-arrive.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+- feedbackDialogResponses = {[
+- { id: 'incorrect', label: 'Incorrect info' },
+- { id: 'missing', label: 'Missing info', `mustComment`: true },
+- { id: 'other', label: 'Other (please specify below)', `mustComment`: true },
+- ]}
+- />
+  ...
+  </Grid>
+  );
```
Comment (kuangp) [inline .changeset/perfect-shoes-arrive.md:22]: Hm you can wrap this in a [Markdown code block](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks) so that it renders nicely in the changelog

File: .changeset/perfect-shoes-arrive.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+<!-- @backstage/plugin-entity-feedback-backend -->
+
+Add in description for 400 response when rating and not authenticated
```
Comment (kuangp) [inline .changeset/perfect-shoes-arrive.md:26]: I recommend probably splitting this into two separate changesets since they are unrelated changes

File: plugins/entity-feedback/README.md [hunk 3]
```diff
@@ -90,6 +90,24 @@ const overviewContent = (
   </Grid>
 );
 
+// Link the feedback comment box to specific feedback responses
```
Comment (kuangp) [inline plugins/entity-feedback/README.md:93]:
```suggestion
// Make feedback comment box required for specific feedback responses
```

**Review by kuangp — CHANGES_REQUESTED** [review]
Thanks for this! Left some thoughts on the UI behavior - let me know what you think

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx [hunk 4]
```diff
@@ -102,13 +136,29 @@ export const FeedbackResponseDialog = (props: FeedbackResponseDialogProps) => {
     }
   }, [comments, consent, entity, feedbackApi, onClose, responseSelections]);
 
+  const selectMandatedBox = (res: boolean) => {
+    const newResponseSelections = { ...responseSelections };
+    requireComments.forEach(id => newResponseSelections[id] === res);
```
[the remark this answers is not in the document]
Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:141]: Hi @kuangp the idea here is to link the comment boxes to specific check boxes. 
- if a user checks a box where mustComment is true, comments must be submitted.
- if a user submits comments when no box is checked, select the tied response. 
This was requested by our design team, but if this is not something that backstage wants to support, please advise on what the behavior should be.

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx [hunk 5]
```diff
@@ -81,13 +87,41 @@ export const FeedbackResponseDialog = (props: FeedbackResponseDialogProps) => {
...
+        alertApi.post({
+          message:
+            'The selected option(s) require a comment. Please provide a comment.',
+          severity: 'info',
+        });
+        return;
+      }
+      if (comments.length > 0 && !isMandatedBoxChecked()) {
+        alertApi.post({
+          message: 'Please select the option(s) that require a comment.',
```
[the remark this answers is not in the document]
Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:119]: Can we leave this as an option to pass in (with `mustComment`) but remove it from the default `feedbackDialogResponses`?

[the remark this answers is not in the document]
Comment (kuangp) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:119]: sure I think a option to enable this works - apart from adding this option, it might be more intuitive that if this is enabled, instead of producing this alert, have the UI disable the comment box unless a comment required option is selected (and ignore the comment if those options are unselected). That way you would never get into this scenario

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (kuangp) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:141]: I think the behavior described in your second point is potentially unintuitive, perhaps you can bring this back to your design team for discussion:

Consider the following example - your feedback dialog has the following options:
  - This entity is deprecated (mustComment)
  - This entity is incorrect (mustComment)
  - Other (mustComment)

Now if the user enters some comments this will auto select all options. However, we don't know which option the user intends to comment on and whether every option is actually applicable. The user may have only wanted to provide generic comments but not intend to state that the entity is either deprecated nor incorrect. A confused user would have to go back and unselect those options wondering why they got auto selected for them.

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx [hunk 6]
```diff
@@ -81,13 +87,41 @@ export const FeedbackResponseDialog = (props: FeedbackResponseDialogProps) => {
...
+        alertApi.post({
+          message:
+            'The selected option(s) require a comment. Please provide a comment.',
+          severity: 'info',
+        });
+        return;
+      }
+      if (comments.length > 0 && !isLinkedBoxChecked()) {
+        alertApi.post({
+          message: 'Please select the option(s) that require a comment.',
```
Comment (vinzscam) [inline plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:119]: perhaps you can move this block outside `useAsyncFn` in another function:

```ts
const handleSaveResponse = () => {
    if (requireComments.length > 0) {
      if (comments.length === 0 && isLinkedBoxChecked()) {
        alertApi.post({
          message:
            'The selected option(s) require a comment. Please provide a comment.',
          severity: 'info',
        });
        return;
      }
      if(....) { }
      saveResponse();
};
```

otherwise you need to add `requireComments`, `alertApi` and `isLinkedBoxChecked` as dependencies (without forgetting to wrap `isLinkedBoxChecked` in a `useMemo`)

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:119]: Hi @kuangp I reviewed your feedback with our design lead & she agreed to make the following changes:
- remove link between comment text and box selection
- remove alerts
- add a comment box for each option where `mustComment` is `true`, this comment box should be hidden until the user checks this box. Filling out the comment box should be optional. 
- the Feedback response table should have some tie between comments & responses
Should I mark this PR as `Draft` until I can get those changes up?

[the remark this answers is not in the document]
Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx:119]: Hi @kuangp just FYI, our design team provided specks with the updated suggestions. I'm including it as a visual to what I described in my comment last week
Screenshot 2024-02-13 at 10 39 46 AM

File: plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx (hunk 6 above)
Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseDialog/FeedbackResponseDialog.tsx]: I removed these changes as a result of feedback received from @kuangp

**Comment by nikolarAutodesk** [comment]
Hey @kuangp I removed the Feedback Dialog changes from this PR. The design changes I was going to make are quite hefty & involve a lot of changes in `FeedbackResponseTable`, so I would prefer to put them in a new PR. In the meantime, I left the smaller changes that we want to contribute.

File: .changeset/perfect-shoes-arrive.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+---
+'@backstage/plugin-entity-feedback': minor
```
Comment (kuangp) [inline .changeset/perfect-shoes-arrive.md:2]: Since this is no longer a breaking change
```suggestion
'@backstage/plugin-entity-feedback': patch
```

File: .changeset/cyan-toes-repeat.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-entity-feedback-backend': patch
+---
+
+Add in description for 400 response when rating and not authenticated
```
Comment (kuangp) [inline .changeset/cyan-toes-repeat.md:5]:
```suggestion
Add in description for 400 response when encountering an invalid rating request
```

**Review by kuangp — CHANGES_REQUESTED** [review]
Some updates to the changesets and we should be good to go  - thanks!

File: plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx [hunk 9]
```diff
@@ -80,9 +80,11 @@ export const FeedbackResponseTable = (props: FeedbackResponseTableProps) => {
       width: '35%',
       render: (response: ResponseRow) => (
         <>
-          {response.response?.split(',').map(res => (
-            <Chip key={res} size="small" label={res} />
-          ))}
+          {response?.response &&
+            response.response.length > 0 &&
+            response.response
```
Comment (nikolarAutodesk) [inline plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx:85]: Previous suggestion to shorten this to: 
```
response.response?.length &&
            response.response
```
led to the following behavior
Screenshot 2024-02-27 at 4 53 17 PM

Comment (kuangp) [reply plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx:85]: hm ok how about `response.response?.length > 0 && response.response` ?

Comment (freben) [reply plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx:85]: What's up with that final `&& response.response`? After a length check. Is it an array or is it not? :)

Comment (freben) [reply plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx:85]: Could we ignore the length check in the first place? It's only gonna be split into an array immediately anyway, and rendering an empty array is fine and does nothing.

`(response.response || '').split(....`

File: .changeset/perfect-shoes-arrive.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+---
+'@backstage/plugin-entity-feedback': patch
+---
+
+<!-- @backstage/plugin-entity-feedback -->
+
```
Comment (freben) [inline .changeset/perfect-shoes-arrive.md:6]: This can be removed btw

File: plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx (hunk 9 above)
Comment (kuangp) [reply plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx:85]: > Oh and as a general note. If you get a zero rendered, it's almost guaranteed to be a case of code on the form
> 
> ```tsx
> {condition && <Component />}
> ```
> 
> which is dangerous since _anything_ falsy will be sent for rendering (the `&&` evaluates to whatever was on the left side of the operator). And a zero is falsy but React happily renders it as a zero to the browser. This can generally be rewritten to
> 
> ```tsx
> {condition ? <Component /> : null}
> ```
> 
> to ensure that a null is what gets "rendered", not the outcome of the condition

ah TIL!

Comment (nikolarAutodesk) [reply plugins/entity-feedback/src/components/FeedbackResponseTable/FeedbackResponseTable.tsx]: That suggestion works as expected @freben, thank you! 
Screenshot 2024-02-28 at 9 57 41 AM
Screenshot 2024-02-28 at 9 57 30 AM

**Review by freben — APPROVED** [review]
